builtin/credential/aws/client.go

```diff
@@ -80,20 +81,39 @@ func (b *backend) getClientConfig(s logical.Storage, region, stsRole, clientType
 		return nil, fmt.Errorf("could not compile valid credentials through the default provider chain")
 	}
 
+	stsConfig, err := b.getRawClientConfig(s, region, "sts")
```

Comment:
Good question (and I probably should have commented it).

Previously, Vault just assumed that if there wasn't an `stsRole` configured for the given account, to just use the default credentials provided. That worked because you were looking up two things:

1. Instance IDs, which were basically guaranteed to be globally unique, so if you found the given instance ID, you knew it was **the** instance ID you were looking for.
1. IAM instance profiles by name. These weren't globally unique across accounts, but it was OK because the only thing you looked for was the associated IAM role ARN, which does include the account number, so you were OK if you looked up the profile in the wrong account because it wouldn't have the associated IAM ARN you need.

In this PR, I need to resolve unique IDs, which requires looking up a particular IAM entity (user or role) in a given account with a given name. If I have an `stsRole` configured for the given account, then great, I'll just try to assume it. But, what if I don't have an `stsRole` defined for that account ID? I'm not a huge fan of the assumption that is being made that it should just look up using the default credentials -- I'm looking up the entity by name, but entities with the same name can exist in multiple accounts. I could check the returned ARN to ensure it has the expected account ID embedded in it, but this feels more correct to me. `getClientConfig` is being asked for an AWS API client in account 123, we should check to ensure that the client we're returning is actually in account 123. And that's the goal here.

Thinking through this a little more, there's a small issue here in that an operator could call `auth/aws/config/sts` with an `account_id` of 123456789012, but pass in an `stsRole` belonging to a different account, as the `config/sts` endpoint doesn't actually validate that the passed-in `stsRole` ARN belongs to `account_id` but I think that's probably a separate discussion (and I'll open a separate issue for it).

> Will it always be a valid assumption to assume that you're getting STS credentials?

I'd put it a little differently. I believe it is a valid assumption that, if you can make *any* AWS API call, you are able to call `sts:GetCallerIdentity`. I've even tried an IAM policy to explicitly **deny** sts:GetCallerIdentity, and it still succeeds.

Does this all make sense?